### ISSUE TITLE
Bundled SDL2 library on Linux does not recognize gyro

### ISSUE THREAD

**Issue by JustinWayland** [Helion-Engine/Helion#882]
Using a Dualshock 4 controller on Linux.  Can confirm this is only an issue on Helion, as Woof! works fine.

Well, unless the gyro needs to be bound to the Turn Left and Turn Right actions first.  In that case, directions on what to do would be appreciated.

**Comment by lemming104** [comment]
Gyro should be enabled by default on controllers that support it, with the gyro yaw axis bound to left/right turn and the pitch axis bound to look up/down.  If the _rest_ of the controller support is working, then SDL either doesn't think you have a gyro, or isn't _telling_ us you have a gyro in the same way that it does on Windows.  Haven't tested that yet.

**Comment by JustinWayland** [comment]
I am running this from the terminal, so adding debug messages would be nice.

Also, looking through the gyro commit, I did not see an equivalent to [this](https://github.com/fabiangreffrath/woof/commit/31d7a394907866134b2cd05ddc392232006aa51a), which could be the reason this is not working.

**Comment by lemming104** [comment]
The actual "talk to SDL" implementation is handled in https://github.com/Helion-Engine/SDLInputWrapper, and the hints are set on init.  https://github.com/Helion-Engine/SDLInputWrapper/blob/e162dc52a41bf26aece1c16b5ed38a1ce76060e3/SDLControllerWrapper/SDLControllerWrapper.cs#L64

**Comment by JustinWayland** [comment]
Then I'm all out of ideas.  Though I did find [this](https://discourse.libsdl.org/t/dualshock-4-gyro-recognized-by-evdev-but-not-sdl/28561/2) on the SDL forums.  Maybe we can't trust the values we get from the sensor update event?

Though it _is_ more than 3 years old.  [This](https://github.com/libsdl-org/SDL/blob/419ae29d90d3934e1a382cd693ac60a1f492e43a/test/testgamecontroller.c) might be a better reference.

It could also be being reported under the accelerometer sensor, maybe?

I am willing to be a guinea pig whose executable spams out controller event debug messages.

**Comment by lemming104** [comment]
I've got another machine with Linux (Ubuntu 24.x) and a DS4 controller to test it with.  On Windows, SDL_GameControllerHasSensor() reports "true" for the accelerometer and gyroscope sensor types, and false for all other types.  On Linux, with the same controller and the same version of SDL2 (the latest stable release), it reports "false" for all types.  Forcibly enabling the sensors anyway has no effect.

**Comment by JustinWayland** [comment]
OK, that's really weird.  Woof! also uses SDL, and it doesn't have any of those problems.

Are the following udev rules in play?  I got these from Steam's installed udev rules, and apparently you need these in order for motion controls to work, otherwise the Dualshock 4 is treated as an XInput device and gyro is disabled.

```
# DualShock 4 over USB hidraw
KERNEL=="hidraw*", ATTRS{idVendor}=="054c", ATTRS{idProduct}=="05c4", MODE="0660", TAG+="uaccess"

# DualShock 4 wireless adapter over USB hidraw
KERNEL=="hidraw*", ATTRS{idVendor}=="054c", ATTRS{idProduct}=="0ba0", MODE="0660", TAG+="uaccess"

# DualShock 4 Slim over USB hidraw
KERNEL=="hidraw*", ATTRS{idVendor}=="054c", ATTRS{idProduct}=="09cc", MODE="0660", TAG+="uaccess"

# DualShock 4 over bluetooth hidraw
KERNEL=="hidraw*", KERNELS=="*054C:05C4*", MODE="0660", TAG+="uaccess"
```

Although you could also just install Steam if you can't be bothered to mess around with udev rules.

(I know this because I switched from Tumbleweed to CachyOS, and gyro was not working on Woof! until I installed Steam, at which point Woof! could detect the gyro on my controller.)

**Comment by JustinWayland** [comment]
Also, I looked at Woof!'s source code, since it seems to be working on my machine.  The input loop is kind of weird and different from what we are doing, but Woof! does this when reading from the Gyro:

```c

void I_ReadGyro(void)
{
    if (I_GyroEnabled() && gyro_supported)
    {
        static event_t ev = {.type = ev_gyro};
        static float data[3];

        SDL_GameControllerGetSensorData(gamepad, SDL_SENSOR_ACCEL, data, 3);
        data[0] /= SDL_STANDARD_GRAVITY;
        data[1] /= SDL_STANDARD_GRAVITY;
        data[2] /= SDL_STANDARD_GRAVITY;
        I_UpdateAccelData(data);

        SDL_GameControllerGetSensorData(gamepad, SDL_SENSOR_GYRO, data, 3);
        ev.data1.f = data[0];
        ev.data2.f = data[1];
        ev.data3.f = data[2];
        D_PostEvent(&ev);
    }
}
```

So instead of getting it's values from the sensor update event it's just reading from the gyro sensor directly.

That being said, I'm not sure how applicable this is to Helion's implementation.  Sorry if I'm not being much help, I'm just puzzled why Woof!'s implementation worked on Linux while Helion's is struggling.

**Comment by lemming104** [comment]
It appears that this might be an OS versioning thing.  The version of libSDL2 I was distributing was built on Ubuntu 22.04.  If I rebuild it on Ubuntu 24.04, it finds the gyros.

I hate developing for Linux.

**Comment by JustinWayland** [comment]
As a casual programmer using Linux, that entire comment made me wince in sympathy, if it makes you feel any better.

**Comment by lemming104** [comment]
I am going to have to think about what this means for our ability to distribute binaries.  We had the inverse of this problem with ZMusic and its deps--I tried building that on 24.04 and distributing a binary based on that, and then users of some random version of Linux Mint based on 22.04 complained that they had no sound.

I may have to maintain binaries for the native deps of various different Linux flavors.  Probably at least the last two LTS versions of Ubuntu, since Ubuntu-based distros (Kubuntu, Mint, etc.) seem to be very popular.

**Comment by JustinWayland** [comment]
Oh that's nasty.

Given that this seems to be a versioning issue, I'll do a build on my system and see if that fixes this.

**Comment by lemming104** [comment]
Well, we're _distributing_ a version of libSDL2 with the application, so you'd need to build and supply your own .so file (drop it in the same directory as Helion) for that.

**Comment by JustinWayland** [comment]
That helped finally triangulate the issue.

Renaming "libSDL2.so" to "libSDL2.so.bak" in the Helion output directory fixes the issue by forcing Helion to use the version of SDL2 installed from the CachyOS repositories, and now gyro input works properly.  Played through E1M1 with only two significant issues, both mostly QoL:
- Missing gyro button, already reported by me.
- Gyro did not respect the mouse look setting.  Setting the vertical sensitivity to zero worked around the issue.

Now I'm wishing that there was a way to have Helion use the system libraries instead on a case-by-case basis.

I _am_ on an Arch-based system, meaning I have the latest versions of the libraries in the repositories. If you want to see how many packaged libraries can be safely dropped from the distributed version without causing significant issues, I can now do that quickly and easily.

I can also try to help figure out what to do about distributing Helion on Linux systems.

**Comment by lemming104** [comment]
I'll probably need to go back through all of our native deps and make sure that the wrappers around them will all _accept_ the system libs.  Basically, when you're using .NET wrappers over native deps, there's some stuff you sometimes need to do to determine _where_ it's going to look for that stuff if you want to override default behaviors.

In terms of native dependencies that we _distribute_, we currently have these:  https://github.com/Helion-Engine/Helion/tree/dev/Client/Unmanaged/binary/linux-x64
- libsdl2
- libfluidsynth (3)
- libzmusic

Plus these, which come in through other package dependencies:
- libglfw(3.3)

We expect these to be installed on the system:  
- Some version of OpenAL
- libsndfile and libmpg123 are implied by libfluidsynth and libzmusic

None of the stuff we presently distribute in binary form is installed by default on Ubuntu.  I don't know about other popular Linuxes.  I don't believe libzmusic is available through any package management source, but I could be wrong.

**Comment by JustinWayland** [comment]
ZMusic is available in the Arch User Repository, but that's an external repo (albeit an extremely popular and officially endorsed one).  For the most part, you are right that ZMusic is not available through any package management source.

SDL2 and fluidsynth are relatively common.

**Comment by lemming104** [comment]
Interesting.  Seems like it's more a question of the _provenance_ of the SDL2 lib than its actual version.  In https://github.com/Helion-Engine/Helion/pull/889, I replaced the bundled libSDL2.so (which I built from source on Ubuntu WSL 22.04) with one ripped from the Ubuntu 22.04 sdl2 dev package, and this one seems to work with full gyro and rumble support on my 24.04 "real" OS.

If you're able to confirm (there should be a dev build that snapshots overnight), then I think we can close this out for now, although I am going to make a note to ensure that Linux stuff should "live off the land" (or, more accurately, out of /usr/lib) if somebody strips out most or all of our prepackaged .so files and tries to use whatever is installed on their system.
